File: algorithm/95. Unique Binary Search Trees II/95.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        if(n==0)    return vector<TreeNode*>();
        return helper(1,n);
    }
private:
    vector<TreeNode*> helper(int start, int end) {
        vector<TreeNode*>   list;
        
        if(start > end) {
            list.push_back(NULL);
            return list;
        }
        
        if(start == end) {
            list.push_back(new TreeNode(start));
            return list;
        }
        
        vector<TreeNode*>   left;
        vector<TreeNode*>   right;

        for(int i = start; i <= end; i++) {
            left = helper(start, i-1);
            right = helper(i+1, end);
            
            for(auto lnode:left) {
                for(auto rnode:right) {
                    TreeNode* cur = new TreeNode(i);
                    cur->left = lnode;
                    cur->right = rnode;
                    list.push_back(cur);
                }
            }
        }
        return list;
        
    }
    
};

File: algorithm/95. Unique Binary Search Trees II/95.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        if(n == 0)  return new ArrayList<TreeNode>();
        return helper(1,n);
    }
    
    private List<TreeNode>  helper(int start, int end) {
        List<TreeNode> list = new ArrayList<TreeNode>();
        
        if(start > end) {
            list.add(null);
            return list;
        }
        
        if(start == end) {
            list.add(new TreeNode(start));
            return list;
        }
        List<TreeNode> left = new ArrayList<TreeNode>();
        List<TreeNode> right = new ArrayList<TreeNode>();
        
        for(int i = start; i <= end; i++) {
            left = helper(start, i - 1);
            right = helper(i + 1, end);
            for(TreeNode lnode:left) {
                for(TreeNode rnode:right) {
                    TreeNode cur = new TreeNode(i);
                    cur.left = lnode;
                    cur.right = rnode;
                    list.add(cur);
                }
            }
        }
        return list;
    }
}
